### ISSUE TITLE
HyperTTS crashes on generation in browser

### ISSUE THREAD

**Issue by herrsilen** [Language-Tools/anki-hyper-tts#137]
Anki crashes when I try to generate audio with HyperTTS through the browser window. However, when generating audio in the Add window, it works.

No error message is displayed.

Steps to recreate:

1. Open card in browser.
2. Fill out fields.
3. Click HyperTTS: Add audio.

**Comment by Tosakun** [comment]
Ditto... every time I've tried generating new audio in the browser field, it crashes.

**Comment by luc-languagetools** [comment]
Anki version and HyperTTS version please

**Comment by Tosakun** [comment]
Anki Version ⁨23.10

HyperTTS 1.01

**Comment by herrsilen** [comment]
Same here:

Anki Version ⁨23.10

HyperTTS 1.01

**Comment by luc-languagetools** [comment]
Is it a hard crash or is there a stack trace (anki pop-up)?

**Comment by herrsilen** [comment]
It is a hard crash. I have noticed that if I generate audio for many cards at the same time, the crash doesn't happen. But if it's only one card, it's every time.

**Comment by Tosakun** [comment]
Been fiddling around with mine.  Awesome TTS (v1.82) still works for me, whether it's individual cards, bulk add, or in the card creation screen.  HyperTTS however when adding audio with the buttons for an individual card or in the card creation screen, it doesn't generate any audio; it doesn't crash either.  .  But every time I try to bulk add audio with HyperTTS, Anki hard crashes with no stack trace.

**Comment by Tosakun** [comment]
Just updated to the newest Anki, 23.10.01, and that seems to have fixed it.

**Comment by Tosakun** [comment]
Never mind, it only made it works if multiple cards are getting audio generate at once.

**Comment by luc-languagetools** [comment]
@Tosakun @herrsilen which operating system ? Any way you could record a video ? I'm having a hard time reproducing this.

**Comment by herrsilen** [comment]
MacOS Sonoma 14.0.

Here's a video: https://share.cleanshot.com/SCL551FS

**Comment by luc-languagetools** [comment]
perfect thank you so much. I think it might be time for me to setup a mac virtual machine to try to reproduce this. I can't reproduce on windows / linux, so maybe there's something related to macs.

**Comment by herrsilen** [comment]
Thank you for looking into it! Let me know if I can help out in any way.

**Comment by kamitchell** [comment]
I'm having this problem on macOS Sonoma. I ran 

```
env AWESOMETTS_DEBUG_LOGGING=1 /Applications/Anki.app/Contents/MacOS/anki
```

When I clicked the button to add audio, the app crashed. I got a crash. I included the entire application log. (click the triangle).

<details><summary>Log</summary>
<p>

```
Anki starting...
Initial setup...
Preparing to run...
Starting main loop...
2023-11-11 09:46:49.324 anki[12079:3024145] WARNING: Secure coding is automatically enabled for restorable state! However, not on all supported macOS versions of this application. Opt-in to secure coding explicitly by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState:.
conf key heatmap should be fetched with col.get_config(), and saved with col.set_config()
conf key heatmap should be fetched with col.get_config(), and saved with col.set_config()
conf key heatmap should be fetched with col.get_config(), and saved with col.set_config()
Qt warning: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=225.891,221.262 gbl=225.891,221.262 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-225.891,-221.262 last=-225.891,-221.262 Δ 225.891,221.262) : no target window
Qt warning: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=301.777,81.878 gbl=301.777,81.878 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-301.777,-81.878 last=-301.777,-81.878 Δ 301.777,81.878) : no target window
Qt warning: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=333.708,90.9165 gbl=333.708,90.9165 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-333.708,-90.9165 last=-333.708,-90.9165 Δ 333.708,90.9165) : no target window
2023-11-11 09:46:52.598 anki[12079:3024145] TSMSendMessageToUIServer: CFMessagePortSendRequest FAILED(-1) to send to port com.apple.tsm.uiserver
2023-11-11 09:46:52.598 anki[12079:3024145] TSMSendMessageToUIServer: CFMessagePortSendRequest FAILED(-1) to send to port com.apple.tsm.uiserver
2023-11-11 09:46:52.599 anki[12079:3024145] TSMSendMessageToUIServer: CFMessagePortSendRequest FAILED(-1) to send to port com.apple.tsm.uiserver
2023-11-11 09:46:52.599 anki[12079:3024145] TSMSendMessageToUIServer: CFMessagePortSendRequest FAILED(-1) to send to port com.apple.tsm.uiserver
JS warning /_anki/js/editor.js:438 No version information available for component [tex]/noerrors
JS warning /_anki/js/editor.js:438 No version information available for component [tex]/mathtools
JS warning /_anki/js/editor.js:438 No version information available for component [tex]/mhchem
JS warning /_anki/js/editor.js:530 The specified value "#00f" does not conform to the required format.  The format is "#rrggbb" where rr, gg, bb are two-digit hexadecimal numbers.
JS warning /_anki/js/editor.js:530 The specified value "#00f" does not conform to the required format.  The format is "#rrggbb" where rr, gg, bb are two-digit hexadecimal numbers.
2023-11-11 09:46:59.122 anki[12079:3024145] TSM AdjustCapsLockLEDForKeyTransitionHandling - _ISSetPhysicalKeyboardCapsLockLED Inhibit
Qt warning: QObject::startTimer: Timers cannot be started from another thread
Qt warning: QObject::startTimer: Timers cannot be started from another thread
Qt warning: QObject::startTimer: Timers cannot be started from another thread
Qt warning: QObject::startTimer: Timers cannot be started from another thread
2023-11-11 09:47:42.250 anki[12079:3024313] *** Assertion failure in -[QCocoaNSMenu _lockForMainMenuItemArray], NSMenu.m:1097
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'API misuse: modification of a menu's items on a non-main thread when the menu is part of the main menu. Main menu contents may only be modified from the main thread.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001824ce800 __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x0000000181fc5eb4 objc_exception_throw + 60
	2   Foundation                          0x000000018361547c -[NSCalendarDate initWithCoder:] + 0
	3   AppKit                              0x0000000185c25440 -[NSMenu itemArray] + 32
	4   libqcocoa.dylib                     0x00000001218a8508 qt_plugin_instance + 90188
	5   QtWidgets                           0x000000012086104c _ZN5QMenu11actionEventEP12QActionEvent + 1228
	6   QtWidgets.abi3.so                   0x000000012131b324 _ZN8sipQMenu11actionEventEP12QActionEvent + 148
	7   QtWidgets                           0x000000012070c6b8 _ZN7QWidget5eventEP6QEvent + 132
	8   QtWidgets                           0x000000012085ef88 _ZN5QMenu5eventEP6QEvent + 1228
	9   QtWidgets.abi3.so                   0x0000000121319ebc _ZN8sipQMenu5eventEP6QEvent + 224
	10  QtWidgets                           0x00000001206c32f8 _ZN19QApplicationPrivate13notify_helperEP7QObjectP6QEvent + 272
	11  QtWidgets                           0x00000001206c4ca0 _ZN12QApplication6notifyEP7QObjectP6QEvent + 3396
	12  QtWidgets.abi3.so                   0x00000001213d7c24 _ZN15sipQApplication6notifyEP7QObjectP6QEvent + 248
	13  QtCore                              0x000000011975b06c _ZN16QCoreApplication15notifyInternal2EP7QObjectP6QEvent + 292
	14  QtGui                               0x000000011cd0b114 _ZN7QAction5eventEP6QEvent + 68
	15  QtGui.abi3.so                       0x000000011bf4a844 _ZN10sipQAction5eventEP6QEvent + 224
	16  QtWidgets                           0x00000001206c32f8 _ZN19QApplicationPrivate13notify_helperEP7QObjectP6QEvent + 272
	17  QtWidgets                           0x00000001206c415c _ZN12QApplication6notifyEP7QObjectP6QEvent + 512
	18  QtWidgets.abi3.so                   0x00000001213d7c24 _ZN15sipQApplication6notifyEP7QObjectP6QEvent + 248
	19  QtCore                              0x000000011975b06c _ZN16QCoreApplication15notifyInternal2EP7QObjectP6QEvent + 292
	20  QtGui                               0x000000011cd0a0b0 _ZN7QAction7setTextERK7QString + 108
	21  QtGui.abi3.so                       0x000000011c01c9dc _ZL20meth_QAction_setTextP7_objectS0_ + 112
	22  anki                                0x000000010237f0c0 cfunction_call + 104
	23  anki                                0x000000010233861c _PyObject_MakeTpCall + 136
	24  anki                                0x0000000102438ea8 _PyEval_EvalFrameDefault + 23972
	25  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	26  anki                                0x000000010233bc30 method_vectorcall + 128
	27  anki                                0x0000000102437134 _PyEval_EvalFrameDefault + 16432
	28  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	29  anki                                0x0000000102339af8 _PyObject_Call_Prepend + 160
	30  anki                                0x00000001023a4364 slot_tp_call + 104
	31  anki                                0x000000010233861c _PyObject_MakeTpCall + 136
	32  anki                                0x0000000102438ea8 _PyEval_EvalFrameDefault + 23972
	33  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	34  anki                                0x000000010233bc30 method_vectorcall + 128
	35  anki                                0x0000000102438024 _PyEval_EvalFrameDefault + 20256
	36  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	37  anki                                0x0000000102437654 _PyEval_EvalFrameDefault + 17744
	38  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	39  anki                                0x0000000102437654 _PyEval_EvalFrameDefault + 17744
	40  anki                                0x000000010243c4e8 _PyEval_EvalCode + 532
	41  anki                                0x0000000102339770 _PyFunction_Vectorcall + 412
	42  anki                                0x0000000102433668 _PyEval_EvalFrameDefault + 1380
	43  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	44  anki                                0x0000000102437654 _PyEval_EvalFrameDefault + 17744
	45  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	46  anki                                0x0000000102433668 _PyEval_EvalFrameDefault + 1380
	47  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	48  anki                                0x0000000102433668 _PyEval_EvalFrameDefault + 1380
	49  anki                                0x000000010243c4e8 _PyEval_EvalCode + 532
	50  anki                                0x0000000102339770 _PyFunction_Vectorcall + 412
	51  anki                                0x0000000102438024 _PyEval_EvalFrameDefault + 20256
	52  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	53  anki                                0x0000000102437654 _PyEval_EvalFrameDefault + 17744
	54  anki                                0x00000001023396c4 _PyFunction_Vectorcall + 240
	55  anki                                0x000000010233bd48 method_vectorcall + 408
	56  anki                                0x0000000102249cc4 t_bootstrap + 72
	57  anki                                0x0000000102492938 pythread_wrapper + 28
	58  libsystem_pthread.dylib             0x000000018237d034 _pthread_start + 136
	59  libsystem_pthread.dylib             0x0000000182377e3c thread_start + 8
)
libc++abi: terminating due to uncaught exception of type NSException
[1]    12079 abort      env AWESOMETTS_DEBUG_LOGGING=1 /Applications/Anki.app/Contents/MacOS/anki
```

</p>
</details> 

Hope this helps track down the problem.

**Comment by jbvoigt3** [comment]
I've had this problem, too, when I run HyperTTS on my Apple, so I've just given up and only run it on my Windows.  Glad someone has raised the issue.Jack(Sent from my iPad)On Nov 11, 2023, at 6:40 PM, Luc ***@***.***> wrote:﻿
perfect thank you so much. I think it might be time for me to setup a mac virtual machine to try to reproduce this. I can't reproduce on windows / linux, so maybe there's something related to macs.

**Comment by luc-languagetools** [comment]
I'm working on getting access to a Mac machine so I can debug this, I appreciate your patience while I work on this.

**Comment by jbvoigt3** [comment]
Thanks. Jack"We are free to choose, but we are not free from the consequences of our choice."On Nov 12, 2023, at 13:04, Luc ***@***.***> wrote:﻿
I'm working on getting access to a Mac machine so I can debug this, I appreciate your patience while I work on this.

**Comment by luc-languagetools** [comment]
Could you all try HyperTTS 1.02 and see whether the crash still happens on MacOS ?

**Comment by jbvoigt3** [comment]
Will do.  I'll get back to you this morning.  Looking forward to the fix.Jack(Sent from my iPad)On Nov 13, 2023, at 9:57 AM, Luc ***@***.***> wrote:﻿
Could you all try HyperTTS 1.02 and see whether the crash still happens on MacOS ?

**Comment by herrsilen** [comment]
> Could you all try HyperTTS 1.02 and see whether the crash still happens on MacOS ?

It works for me. Thanks a million!

**Comment by jbvoigt3** [comment]
Here’s my error message.  BTW: someone mentioned that it would be nice to have the “selection” option back again.  I agree.





Error
An error occurred. Please start Anki while holding down the shift key, which will temporarily disable the add-ons you have installed.
If the issue only occurs when add-ons are enabled, please use the Tools > Add-ons menu item to disable some add-ons and restart Anki, repeating until you discover the add-on that is causing the problem.
When you've discovered the add-on that is causing the problem, please report the issue to the add-on author.
Debug info:
Anki 2.1.66 (70506aeb) Python 3.9.15 Qt 6.5.0 PyQt 6.5.0
Platform: macOS-12.7.1-x86_64-i386-64bit
Flags: frz=True ao=True sv=3
Add-ons, last update check: 2023-11-12 11:30:55
Add-ons possibly involved: ⁨HyperTTS - Add speech to your flashcards⁩

Caught exception:
Traceback (most recent call last):
  File "aqt.taskman", line 122, in _on_closures_pending
  File "/Users/jbvoigt/Library/Application Support/Anki2/addons21/111623432/anki_utils.py", line 109, in undo_end_fn
    aqt.mw.col.merge_undo_entries(undo_id)
  File "anki.collection", line 1059, in merge_undo_entries
  File "anki._backend_generated", line 170, in merge_undo_entries
  File "anki._backend", line 151, in _run_command
anki.errors.InvalidInput: target undo op not found

**Comment by luc-languagetools** [comment]
@jbvoigt3 was audio successfully generated for any note in this case ? Or was there an error for one note and actually nothing was updated ?

**Comment by jbvoigt3** [comment]
Audio was generated and file created.  Then error message popped up.

Respectfully,

Jack Voigt
***@***.***

"We are free to choose, but we are not free from the consequences of our
choice."

**Comment by kamitchell** [comment]
> Could you all try HyperTTS 1.02 and see whether the crash still happens on MacOS ?

Fixed the crash when generating audio when browsing or in batches on my macOS machine. Thanks so much!

**Comment by luc-languagetools** [comment]
@kamitchell @herrsilen if the issue is solved for you, can I ask you to write a review ? https://ankiweb.net/shared/info/111623432

**Comment by luc-languagetools** [comment]
@jbvoigt3 could you record a video for me ? I'm trying to reproduce this bug and unable to do so.

**Comment by jbvoigt3** [comment]
Yeah, I can do that.  Let me wake up a little.

Respectfully,

Jack Voigt
***@***.***

"We are free to choose, but we are not free from the consequences of our
choice."

**Comment by jbvoigt3** [comment]
Size makes me think the videos didn’t come through via email.  Here are links from my shared Google drive:

https://drive.google.com/file/d/107ZCuOTyhxW7gj1acSTgX1PzQ7gdxaYa/view?usp=drive_link

https://drive.google.com/file/d/107cKe7xu1m3EZWE9BXMW5TtvG4g5Iyfa/view?usp=drive_link

**Comment by luc-languagetools** [comment]
@jbvoigt3 I don't see the error message `anki.errors.InvalidInput: target undo op not found` appear in your videos, or did I miss it ?

**Comment by jbvoigt3** [comment]
You might have missed it.  I'll redo it.  Sorry, but it will be tomorrow.
Traveling today.

Respectfully,

Jack Voigt
***@***.***

"We are free to choose, but we are not free from the consequences of our
choice."

**Comment by luc-languagetools** [comment]
@jbvoigt3 no need, I was able to reproduce the issue.

**Comment by luc-languagetools** [comment]
@jbvoigt3 it should be fixed in HyperTTS 1.03

**Comment by luc-languagetools** [comment]
I consider that the hard crash on MacOSX is fixed, please let me know otherwise.
